interfaces/builtin/content.go

```diff
@@ -0,0 +1,160 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"bytes"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+)
+
+// ContentInterface allows sharing content between snaps
+type ContentInterface struct{}
+
+func (iface *ContentInterface) Name() string {
+	return "content"
+}
+
+func (iface *ContentInterface) SanitizeSlot(slot *interfaces.Slot) error {
+	if iface.Name() != slot.Interface {
+		panic(fmt.Sprintf("slot is not of interface %q", iface))
+	}
+
+	// check that we have either a read or write path
+	rpath, rok := slot.Attrs["read"].([]interface{})
+	wpath, wok := slot.Attrs["write"].([]interface{})
+	if !rok && !wok {
+		return fmt.Errorf("content must contain a read or write path")
+	}
+	if len(rpath) == 0 && len(wpath) == 0 {
+		return fmt.Errorf("read or write path must be set")
+	}
+
+	// go over both paths
+	paths := rpath
+	paths = append(paths, wpath...)
+	for _, p := range paths {
+		if strings.Contains(p.(string), "..") {
```

Comment:
We need to check if it's a string here as that's a runtime panic otherwise.

Also, a more comprehensive check would be filepath.Clean(p) == p. This ensures the path provided is completely clean, which seems sane.